alf/algorithms/generator.py

```diff
@@ -36,20 +36,41 @@ class Generator(Algorithm):
 
     Generator generates outputs given `inputs` (can be None) by transforming
     a random noise and input using `net`:
-        outputs = net([noise, inputs]) if input is not None
+
+        outputs = net([noise, input]) if input is not None
                   else net(noise)
 
-    It has two modes:
-    ML: net is trained to minized loss_func([outputs, inputs])
-    STEIN: net is trained to generate outputs to match the distribution whose
-        (unnormalized) negative probability is given by loss_func([outputs, inputs]).
-        The matching is achieved by using amortized Stein variational gradient
-        descent (SVGD). See the following paper for reference
-        Feng et al "Learning to Draw Samples with Amortized Stein Variational
-        Gradient Descent" https://arxiv.org/pdf/1707.06626.pdf
+    The generator is trained to minimize the following objective:
+
+        E(loss_func(net([noise, input]))) - entropy_regulariztion * H(P)
+
+    where P is the (conditional) distribution of outputs given the inputs
+    implied by `net` and H(P) is the (conditional) entropy of P.
+
+    If the loss is the (unnormalized) negative log probability of some
+    distribution Q and the entropy_regularization is 1, this objective is
+    equivalent to minimizing KL(P||Q).
+
+    It uses two different ways to optimize `net` depending on
+    entropy_regularization:
+
+    * entropy_regularization = 0: the minimization is achieved by simply
+      minimizing loss_func(net([noise, inputs]))
+
+    * entropy_regularization > 0: the minimization is achieved using amortized
+      Stein variational gradient descent (SVGD). See the following paper for
+      reference:
+
+      Feng et al "Learning to Draw Samples with Amortized Stein Variational
+      Gradient Descent" https://arxiv.org/pdf/1707.06626.pdf
 
```

Comment:
It is a clean way to support both regularized and non-regularized training, and entropy-based regularization does include a large variations of regularization methods. However, I was wondering if we might possibly go beyond entropy regularization in the future? If so, might we want a more abstract generator class?

Reply:
We can do that when there is need. Otherwise if writing a very general generator now, it's probably won't satisfy the new need when it actually comes.